Create documentation for internal workings of PyDough (#336)
bodo-ai/PyDough: documentation/ir.md

2. When `to_sql` or `to_df` is called on this unqualified nodes, it is first sent through a process called [qualification](#qualification) which converts unqualified nodes into qualified DAG nodes, or [QDAG nodes](#qdag) for short. These QDAG nodes utilize the metadata in order to correctly associate every aspect of the PyDough logic with the data being analyzed/transformed, and is also where a great deal of the verification of the PyDough code's validity happens.
3. Next, the QDAG nodes are run through a process called [hybrid conversion](#hybrid-conversion) which restructures the logic into the datastructure known as the [hybrid tree](#hybrid-tree) in order to better organize the types of ways different subtrees of the data are linked together.
4. The hybrid tree is further transformed by the [decorrelation procedure](#hybrid-decorrelation) to remove correlated references created by the hybrid conversion process.

> 5. The transformed hybrid tree is converted into a [relational tree](#relational-tree) highly reminiscent of the datastructure used to represent relational algebra in frameworks such as Apache Calcite. The conversion to this datastructure is called [relational conversion](#relational-conversion).

Author: It is used more as a macrocosm example of what kind of datastructure it is, since this analogy would be more familiar to a lot of people with some database experience when explaining how PyDough works. But yeah I can link it.

  region_name=_ROOT.region.name,
  nation_name=_ROOT.name,
  n_orders_from_debt_customers=_ROOT.COUNT(_ROOT.customers.WHERE(_ROOT.acctbal < 0).orders)
).ORDER_BY(_ROOT.nation_name.ASC())
```

All of these `_ROOT.x` terms become `UnqualifiedAccess` nodes accessing term `x` and pointing to `_ROOT` as the thing they are accessing from, except for `_ROOT.COUNT`, which becomes a function call because the transformation recognizes this as a function name.

There are several variations of the logic that invoke this rewrite in different contexts, (e.g. unit tests vs jupyter notebooks) but they all occur in [unqualified_transform.py](../PyDough/unqualified/unqualified_transform.py).

<!-- TOC --><a name="qdag-nodes"></a>
## QDAG Nodes

> [!IMPORTANT]
> TODO: FINISH THIS SECTION.

For an example of the QDAG structure, consider the `nation_info` example earlier from the unqualified nodes:

```py
nation_info = nations.CALCULATE(
  region_name=region.name,
  nation_name=name,
  n_orders_from_debt_customers=COUNT(customers.WHERE(acctbal < 0).orders)
).ORDER_BY(nation_name.ASC())
```

This has the following structure as QDAG nodes:

```
──┬─ TPCH
  ├─── TableCollection[nations]

>   ├─┬─ Calculate[region_name=$1.name, nation_name=name, n_customers_in_debt=COUNT($2)]

Author: Whoops, this should have been updated but wasn't.

  │ ├─┬─ AccessChild
  │ │ └─── SubCollection[region]
  │ └─┬─ AccessChild
  │   ├─── SubCollection[customers]
  │   └─┬─ Where[acctbal < 0]
  │     └─── SubCollection[orders]
  └─── OrderBy[nation_name.ASC(na_pos='first')]
```

<!-- TOC --><a name="qualification"></a>
### Qualification

> [!IMPORTANT]
> TODO: FINISH THIS SECTION.

<!-- TOC --><a name="hybrid-tree"></a>
## Hybrid Tree

> [!IMPORTANT]
> TODO: FINISH THIS SECTION.

For an example of the Hybrid Tree structure, consider the `nation_info` example earlier from the unqualified nodes:

```py
nation_info = nations.CALCULATE(
  region_name=region.name,
  nation_name=name,

>   n_customers_in_debt=COUNT(customers.WHERE(acctbal < 0).orders)

Author: Yeah this is out of date and should be updated.

  "]
  end
  L1 --> L2
  L2 -.->|Singular| L3c
  L2 -.->|Aggregation| L4c
  L4 --> L5
```

In this example, the main hybrid tree has two levels `H1` and `H2` (`H2` is the value pointed to when the QDAG is converted to a hybrid tree).
- `H1`/`H2` have a parent/successor relationship where `H1` is the parent and `H2` is the successor.
- `H1` has no children and its pipeline has a single operation denoting the root level.
- `H2` has two children (`$0` and `$1`) and its pipeline has 3 operations:
  - An access to the nations collection (how the step-down from `H1` to `H2` begins)
  - A calculate that defines `region_name`, `nation_name`, and `n_customers_in_debt`
  - An order-by that sorts by `nation_name` in ascending order.
- `$0` is a singular access, meaning the data from `H2` and `$0` can be directly joined without needing to worry about changes in cardinality. They are joined on the condition that the `region_key` term from `H2` equals the `key` term from the bottom subtree of `$0` (which is `H3`).
  - The contents of `$0` is just a single tree level `H3` which does not have any children and has a pipeline containing only an access to the `regions` collection.

> - `$1` is an aggregation access, meaning the data from `$1` must first be aggregated before it is joined with `H2`. The aggregation is done by grouping on the `nation_key` field of the bottom subtree of `$1` (which is `H5`) and computes the term `agg_0` as `COUNT()`. Then, the result is joined with `H2` on the condition that the `key` term from `H2` equals the `nation_key` field just used to aggregate.

Reviewer: Suggestion to make $0 and $1 points have sublists describing steps rather than paragraph.
E.g.:
$1 is an aggregation access over customers.orders, with:
Grouped by nation_key and joined to H2 on key = nation_key
COUNT() aggregation over the orders collection
Join on condition ....

>   - The contents of `$1` is two tree levels `H4` and `H5`. `H4` does not have any children and has a pipeline containing an access to the `customers` collection followed by a filter on the condition that `acctbal < 0`. `H5` does not have any children and only contains a single operation accessing the `orders` sub-collection of the parent level.

Reviewer: ```suggestion
  - The contents of `$1` is two levels `H4` and `H5`. `H4` does not have any children and has a pipeline containing an access to the `customers` collection followed by a filter on the condition that `acctbal < 0`. `H5` does not have any children and only contains a single operation accessing the `orders` sub-collection of the parent level.
```

<!-- TOC --><a name="hybrid-conversion"></a>
### Hybrid Conversion

> [!IMPORTANT]
> TODO: FINISH THIS SECTION.

<!-- TOC --><a name="hybrid-decorrelation"></a>
### Hybrid Decorrelation

> To understand why de-correlation matters, first consider the slightly more complex PyDough code example below. This PyDough code finds, for each nation in Europe, the total purchase quantity made by customers in that nation from suppliers in the same nation. 5 nations with the largest number of orders made by customers in that nation in the building market segment where the total price of the order is at least double the average of the total prices of **all** orders.

Reviewer: ```suggestion
To understand why de-correlation matters, first consider the slightly more complex PyDough code example below. This PyDough code finds, for each nation in Europe, the total quantity of purchases made by customers from suppliers in the same nation. 5 nations with the largest number of orders made by customers in that nation in the building market segment where the total price of the order is at least double the average of the total prices of **all** orders.
```

Reviewer: Is this part of the example? I don't see top 5 in the code
> 5 nations with the largest number of orders made by customers in that nation in the building market segment where the total price of the order is at least double the average of the total prices of **all** orders.

Author: Whoops, more holdover from an older version I forgot to delete.

    name == 'EUROPE'"]
    R --> S
  end

  subgraph L2["(H3)"]
  direction LR
    E["Sub-collection:
    nations"]
    F["Calculate
    nation_name: name"]
    G["Calculate
    nation_name
    domestic_quantity = DEFAULT_TO($0.agg_0, 0)"]
    E --> F --> G
  end

  subgraph C2["($0)"]
    subgraph L4["(H4)"]
    direction LR
      C["Sub-Collection:
      customers"]
    end
    subgraph L5["(H5)"]
    direction LR
      D["Sub-Collection:
      orders"]
    end
    subgraph L6["(H6)"]
    direction LR
      V1["Sub-Collection:
      lines"]
      V2["Where
      $0.name == CORREL(nation_name)"]
      V1 --> V2
    end
    subgraph C4["($0)"]
    direction TB
        subgraph H7["(H7)"]
        Q1["Sub-Collection:
      supplier"]
        end
        subgraph H8["(H8)"]
        Q2["Sub-Collection:
      nation"]
        end
      data2["Child Info:
        join_keys: [(supplier_key, BACK(1).key)]"]
    end
    data1["Child Info:
    join_keys: [(key, BACK(1).nation_key)]
    agg_keys: [BACK(1).nation_key]
    aggs: {'agg_0': SUM(quantity)}"]
  end

    L1 --> R1
    R1 --> L2
    L2 -.->|Aggregation| C2
    L4 --> L5
    L5 --> L6
    L6 -.->|Singular| C4
    Q1 --> Q2
```

> Notice that `H6` contains a filter condition `$0.name == CORREL(nation_name)`. This means that even though `H6` is inside child `$0` of `H3` (not to be confused with child `$0` of `H6`), it still references the `nation_name` field from `H3`. This is a called a correlated reference, because it means that child `$0` of `H3` requires information from `H3` in order to be derived, but the whole point is that the data from child `$0` is calculated then joined onto he data from `H3`, so there is a catch-22.

Reviewer: This was confusing to me. Here's a suggestion to edit. Feel free to update
```suggestion
Notice that `H6` contains a filter condition `$0.name == CORREL(nation_name)`, indicating a correlated reference. Although `H6` is nested within child `$0` of `H3`, it still references the `nation_name` field from `H3`. This is called a correlated reference — child `$0` of `H3` depends on `H3`'s data (`nation_name`) to compute its result. But since this data is meant to be computed before joining with `H3`, this creates a catch-22.
```

> To fix this, we next run the de-correlation procedure. This will recursively traverse the entire tree and look for hybrid nodes that have a correlated child (where the child type is NOT semi/anti join, since those two do allow correlated references). This procedure will reach tree `H3` and see that its child `$0` is correlated and is an `AGGREGATION` connection, and will therefore de-correlate it. It does so by copying the entire hybrid tree so far (`H1`, `H2`, and the first two operators from `H3`) and attaching them to the top of child `$0`, above `H4`. This now means that the `CORREL(nation_name)` term in `H6` can be rephrased into a back-reference to the `nation_name` field in the copied version of `H3` (which is 3 levels above `H6`). This does change the join/aggregation keys when connecting H3 to child `$0`, since now we must connect each unique record of `H3` to the corresponding prepended section of `$0`. This is done by changing the join keys into the uniqueness keys of `H1`/`H2`/`H3`, which is aka `key` from `regions` (`BACK(1).key` from the perspective of `H3` and `BACK(4).key` from the perspective of `H6`) and `key` from `nations` (`key` from the perspective of `H3` and `BACK(3).key` from the perspective of `H6`).

Reviewer: ```suggestion
To fix this, we next run the de-correlation procedure. This procedure recursively traverses the entire tree, looking for hybrid nodes with correlated children (excluding semi and anti joins, which do allow correlated references). This procedure will reach tree `H3` and see that its child `$0` is correlated and is an `AGGREGATION` connection, and will therefore de-correlate it. To do this, it copies the hybrid tree built so far (`H1`, `H2`, and the first two operators of `H3`) and attaches them above child `$0`, specifically before `H4`. This now means that the `CORREL(nation_name)` term in `H6` can be rephrased into a back-reference to the `nation_name` field in the copied version of `H3` (which is 3 levels above `H6`). This changes the join and aggregation keys used when connecting `H3` to child `$0`, as each unique record from `H3` must now align with the prepended section of `$0`. This is done by changing the join keys into the uniqueness keys of `H1`/`H2`/`H3`, the `key` from `regions` (accessed as `BACK(1).key` from`H3` and `BACK(4).key` from `H6`) and the `key` from `nations` (`key` from  `H3` and `BACK(3).key` from `H6`).
```

        name == 'EUROPE'"]
        R' --> S'
    end

    subgraph L2'["(H3')"]
    direction LR
        E'["Sub-collection:
        nations"]
        F'["Calculate
        nation_name: name"]
        E' --> F'
    end
    subgraph L4["(H4)"]
    direction LR
      C["Sub-Collection:
      customers"]
    end
    subgraph L5["(H5)"]
    direction LR
      D["Sub-Collection:
      orders"]
    end
    subgraph L6["(H6)"]
    direction LR
      V1["Sub-Collection:
      lines"]
      V2["Where
      $0.name == BACK(3).nation_name"]
      V1 --> V2
    end
    subgraph C4["($0)"]
    direction TB
        subgraph H7["(H7)"]
        Q1["Sub-Collection:
      supplier"]
        end
        subgraph H8["(H8)"]
        Q2["Sub-Collection:
      nation"]
        end
      data2["Child Info:
        join_keys: [(supplier_key, BACK(1).key)]"]
    end
    data1["Child Info:
    join_keys: [(key, BACK(3).key), (BACK(1).key, BACK(4).key)]
    agg_keys: [BACK(3).key, BACK(4).key]
    aggs: {'agg_0': SUM(quantity)}"]
  end

    L1 --> R1
    R1 --> L2
    L1' --> R1'
    R1' --> L2'
    L2' -->L4
    L2 -.->|Aggregation| C2
    L4 --> L5
    L5 --> L6
    L6 -.->|Singular| C4
    Q1 --> Q2
```

> However, we can go one step further if we make an observation about this hybrid tree. Notice how now, we are computing the logic of `H1`, `H2`, and most of `H3` twice. This must happen if we intend to keep *every* nation, including the ones without any entries of `$0`. However, if we don't intend to keep nations where there are no records of `$0` to join onto, we can modify the original PyDough code as follows:

Reviewer: ```suggestion
However, we can go one step further if we make an observation about this hybrid tree. Notice how now, we are computing the logic of `H1`, `H2`, and most of `H3` twice. This must happen if we intend to keep *every* nation, including the ones without any entries of `$0`. However, if we don't intend to keep nations that lack corresponding `$0` records to join with, we can modify the original PyDough code as follows:
```

```py
selected_nations = regions.WHERE(
  name == "EUROPE"
).nations.CALCULATE(nation_name=name)
domestic_lines = customers.orders.lines.WHERE(supplier.nation.name == nation_name)
nation_domestic_purchase_info = selected_nations
.WHERE(HAS(domestic_lines)).CALCULATE(
    nation_name,
    domestic_quantity=SUM(domestic_lines.quantity),
)
```

> With this modification, the original hybrid tree changes so the access to child `$0` of `H3` is now an `AGGREGATION_ONLY_MATCH` access, meaning we do an inner join instead of a left join. The de-correlation procedure will notice this and, after transforming child `$0` into its de-correlated form, realize it can do an optimization to delete the original `H1`, `H2` and prefix of `H3` since all of the data required from them is accessible inside `$0`. This is done via the hybrid "pull up" node, which specifies that the data comes from the specified child, instead of having existing data that gets joined with data from the child. This will look as follows:

Reviewer: ```suggestion
With this modification, the original hybrid tree is modified so that access to child `$0` of `H3` is now an `AGGREGATION_ONLY_MATCH` access, meaning we do an inner join instead of a left join. The de-correlation procedure will notice this and, after transforming child `$0` into its de-correlated form, recognize an opportunity to optimize by removing the original `H1`, `H2` and prefix of `H3` since all of the data required from them is accessible inside `$0`. This is done via the hybrid "pull up" node, which specifies that the data comes from the specified child, instead of having existing data that gets joined with data from the child. This will look as follows:
```

        regions"]
        S'["Where
        name == 'EUROPE'"]
        R' --> S'
    end

    subgraph L2'["(H3')"]
    direction LR
        E'["Sub-collection:
        nations"]
        F'["Calculate
        nation_name: name"]
        E' --> F'
    end
    subgraph L4["(H4)"]
    direction LR
      C["Sub-Collection:
      customers"]
    end
    subgraph L5["(H5)"]
    direction LR
      D["Sub-Collection:
      orders"]
    end
    subgraph L6["(H6)"]
    direction LR
      V1["Sub-Collection:
      lines"]
      V2["Where
      $0.name == BACK(3).nation_name"]
      V1 --> V2
    end
    subgraph C4["($0)"]
    direction TB
        subgraph H7["(H7)"]
        Q1["Sub-Collection:
      supplier"]
        end
        subgraph H8["(H8)"]
        Q2["Sub-Collection:
      nation"]
        end
      data2["Child Info:
        join_keys: [(supplier_key, BACK(1).key)]"]
    end
    data1["Child Info:
    agg_keys: [BACK(3).key, BACK(4).key]
    aggs: {'agg_0': SUM(quantity)
    'agg_1': ANYTHING(BACK(3).nation_name)}"]
  end

    L1' --> R1'
    R1' --> L2'
    L2' -->L4
    L2 -.->|AggregationOnlyMatch| C2
    L4 --> L5
    L5 --> L6
    L6 -.->|Singular| C4
    Q1 --> Q2
```

> The way to interpret this is that the entirety of child `$0` of H3 is evaluated, from `H1'` to `H6`, then grouped on `BACK(3).key` and `BACK(4).key` to calculate `agg_0` and `agg_1` (renaming the latter to `nation_name`), and that result is passed along from the PullUp node to the calculate in `H3` which builds on it.

Reviewer: ```suggestion
This means the entire child `$0` subtree of `H3` is evaluated, from `H1'` to `H6`, then grouped on `BACK(3).key` and `BACK(4).key` to calculate `agg_0` and `agg_1` (renaming the latter to `nation_name`), and that result is passed along from the PullUp node to the calculate in `H3` which builds on it.
```

<!-- TOC --><a name="relational-tree"></a>
## Relational Tree

> [!IMPORTANT]
> TODO: FINISH THIS SECTION.

> For an example of the relational tree, consider the `nation_info` example from earlier.

Reviewer: ```suggestion
To illustrate the relational tree, consider the `nation_info` example from earlier.
```

```py
nation_info = nations.CALCULATE(
  region_name=region.name,
  nation_name=name,
  n_orders_from_debt_customers=COUNT(customers.WHERE(acctbal < 0).orders)
).ORDER_BY(nation_name.ASC())
```

> Using the hybrid tree from earlier, the following is how the relational tree created by relational conversion could look:

Reviewer: ```suggestion
Using the hybrid tree from earlier, here’s how the relational tree created by the conversion would look:
```
